README
NAME
    Data::Average - Hold Data Set To Calculate Average

SYNOPSIS
      use Data::Average;

      my $data = Data::Average->new;
      $data->add($_) for (1..100);

      print $data->avg; # 55

DESCRIPTION
    Data::Average is a very simple module: You add values to it, and then
    you compute the average using the avg() function.

METHODS
  new()
    Creates a new Data::Average object

  add($value)
    Adds a value to the Data::Average set.

  length()
    Returns the current data set

  avg()
    Returns the average of the entire set

AUTHOR
    Copyright (c) 2006 Daisuke Maki <dmaki@cpan.org> All rights reserved.



Changes
Changes
=======

0.03000
    - Fix illegal division by zero. rt#30715

0.02 - 17 May 2006
    - Fix document
    - Fix data traversal on arguments for Expires.pm

0.01 - 16 May 2006
    - Initial Release 

lib/Data/Average/Bounded.pm
# $Id: Bounded.pm 1 2006-05-16 15:02:55Z daisuke $
#
# Copyright (c) 2006 Daisuke Maki <dmaki@cpan.org>
# All rights reserved.

package Data::Average::Bounded;
use strict;
use base qw(Data::Average);

sub new
{
    my $class = shift;
    my %args  = @_;
    my $self  = $class->SUPER::new(%args);
    $self->{max} = $args{max};

    return $self;
}

sub add
{
    my $self = shift;
    if ($self->{max} <= $self->length) {
        shift @{$self->{data}};
    }
    $self->SUPER::add(@_);
}

1;

__END__

=head1 NAME

Data::Average::Bounded - Data::Average With Bounded

=head1 SYNOPSIS

  use Data::Average::Bounded;

  my $data = Data::Average::Bounded->new(max => 10);
  $data->add($_) for (1..100);

  print $data->avg; # 95 (avg of 90 .. 100)

=head1 DESCRIPTION

Data::Average::Bounded is a bounded version of Data::Average, which keeps
the data size to a predefined size.

=head1 METHODS

=head2 new(max => $max)

Creates a new Data::Average::Bounded object, bounded by $max.

=head2 add($value)

Adds a value to the Data::Average::Bounded set. If the data size exceeds the 
length specified by 'max' given to the constructor, old elements are popped 
out of the set to keep the data size.

$value may either be an object that implements a method 'value()' which
returns a numerical representation of the object, or a simple scalar.

=head2 length()

=head2 avg()

Same as Data::Average.

=head1 AUTHOR

Copyright (c) 2006 Daisuke Maki E<lt>dmaki@cpan.orgE<gt>
All rights reserved.

=cut


lib/Data/Average/Expires.pm
# $Id: Expires.pm 1 2006-05-16 15:02:55Z daisuke $
#
# Copyright (c) 2006 Daisuke Maki <dmaki@cpan.org>
# All rights reserved.

package Data::Average::Expires;
use strict;
use base qw(Data::Average);

sub new
{
    my $class = shift;
    my %args  = @_;
    my $self  = $class->SUPER::new(%args);
    $self->{expires_in} = $args{expires_in};

    return $self;
}

sub expire_items
{
    my $self   = shift;
    my $length = $self->SUPER::length;
    my $i      = 0;
    while ($i < $length) {
        if ($self->{data}->[$i]->expired) {
            splice(@{$self->{data}}, $i, 1);
            $length--;
        } else {
            $i++;
        }
    }
}

sub length
{
    my $self = shift;
    $self->expire_items();
    $self->SUPER::length(@_);
}

sub avg
{
    my $self = shift;
    $self->expire_items();
    $self->SUPER::avg(@_);
}

sub add
{
    my $self = shift;
    $self->expire_items;

    my @args;
    for (@_) {
        if (!ref($_) || ! $_->isa('Data::Average::Expires::Item')) {
            push @args, Data::Average::Expires::Item->new(expires_in => $self->{expires_in}, value => $_);
        }
    }
    $self->SUPER::add(@args);
}

package Data::Average::Expires::Item;
use strict;

sub new
{
    my $class = shift;
    my %args  = @_;
    my $self  = bless {
        expires_in => $args{expires_in},
        value      => $args{value},
        created_on => time()
    }, $class;
    return $self;
}

sub expires_in { $_[0]->{expires_in} }
sub value      { $_[0]->{value} }
sub created_on { $_[0]->{created_on} }
sub expired { $_[0]->{created_on} + $_[0]->{expires_in} < time() }

1;

__END__

=head1 NAME

Data::Average - Hold Data Set To Calculate Average

=head1 SYNOPSIS

  use Data::Average::Expires;

  my $data = Data::Average::Expires->new(expires_in => 10);
  $data->add($_) for (1..100);

  print $data->avg; # 55

  # sleep for more than 10 seconds...
  sleep(15);
  print $data->length; # 0

  $data->add(Data::Average::Expires::Item->new(expires_in => 60, value => 200));

=head1 DESCRIPTION

Data::Average::Expires only takes into account values that haven't expired:
for example, you can track the average value of something for the last 
10 minutes using this module.

=head1 METHODS

=head2 new(expires_in => $expires_in)

Creates a new Data::Average object, with default expire time set to the
value denoted by $expires_in

=head2 add($value)

Adds a value to the Data::Average::Expires set. 

$value may be a scalar or a Data::Average::Expires::Item object. Items
are always checked for expiry before each operation.

=head2 length()

=head2 avg()

Same as Data::Average.

=head1 AUTHOR

Copyright (c) 2006 Daisuke Maki E<lt>dmaki@cpan.orgE<gt>
All rights reserved.

=cut


lib/Data/Average/BoundedExpires.pm
# $Id: BoundedExpires.pm 1 2006-05-16 15:02:55Z daisuke $
#
# Copyright (c) 2006 Daisuke Maki <dmaki@cpan.org>
# All rights reserved.

package Data::Average::BoundedExpires;
use strict;
use base qw(Data::Average::Expires);

sub new
{
    my $class = shift;
    my %args  = @_;
    my $self  = $class->SUPER::new(%args);
    $self->{max}        = $args{max};

    return $self;
}

sub avg
{
    my $self = shift;
    $self->expire_items();
    $self->SUPER::avg(@_);
}

sub add
{
    my $self = shift;
    $self->expire_items;
    if ($self->{max} <= $self->length) {
        shift @{$self->{data}};
    }
    $self->SUPER::add(@_);
}

1;

__END__

=head1 NAME

Data::Average::BoundedExpires - Bounded Data::Average With Expiry

=head1 SYNOPSIS

  use Data::Average::BoundedExpires;

  my $data = Data::Average::BoundedExpires->new(max => 10, expires_in => 120);

=head1 DESCRIPTION

Data::Average::Bounded is a mix of Data::Average::Bounded and
Data::Average::Expires.

=head1 METHODS

=head2 new(max => $max, expires_in => $expires_in)

Creates a new Data::Average::BoundedExpires object

=head2 add($value)

Behaves like a combination of Data::Average::Bounded and Data::Average::Expires.

=head2 length()

=head2 avg()

Same as Data::Average.

=head1 AUTHOR

Copyright (c) 2006 Daisuke Maki E<lt>dmaki@cpan.orgE<gt>
All rights reserved.

=cut


lib/Data/Average.pm
# $Id: Average.pm 1 2006-05-16 15:02:55Z daisuke $
#
# Copyright (c) 2006 Daisuke Maki <dmaki@cpan.org>
# All rights reserved.

package Data::Average;
use strict;
use UNIVERSAL::isa;
use vars qw($VERSION);

BEGIN
{
    $VERSION = '0.02';
}

sub new
{
    my $class = shift;
    my $self  = bless {
        data => []
    }, $class;
    return $self;
}

sub avg
{
    my $self   = shift;
    my $total  = 0;
    my $length = $self->length;
    for my $i (0..$length - 1) {
        my $v = $self->{data}->[$i];
        if (ref($v) && $v->can('value')) {
            $total += $v->value;
        } else {
            $total += $v;
        }
    }

    return $total / $length;
}

sub length { scalar(@{$_[0]->{data}}) }
sub add    { 
    my $self = shift;
    $self->{data}->[ $self->length ] = $_ for @_;
    1;
}

1;

__END__

=head1 NAME

Data::Average - Hold Data Set To Calculate Average

=head1 SYNOPSIS

  use Data::Average;

  my $data = Data::Average->new;
  $data->add($_) for (1..100);

  print $data->avg; # 55

=head1 DESCRIPTION

Data::Average is a very simple module: You add values to it, and then you
compute the average using the avg() function.

=head1 METHODS

=head2 new()

Creates a new Data::Average object

=head2 add($value)

Adds a value to the Data::Average set.

=head2 length()

Returns the current data set

=head2 avg()

Returns the average of the entire set

=head1 AUTHOR

Copyright (c) 2006 Daisuke Maki E<lt>dmaki@cpan.orgE<gt>
All rights reserved.

=cut